core/spatial_map.cpp

```diff
@@ -788,395 +788,382 @@ void SpatialMap::ColorForValue(double p_value, float *p_rgb_ptr)
 	if (kernel.dimensionality_ != 1)
 		EIDOS_TERMINATION << "ERROR (SpatialMap::Convolve_S1): (internal error) kernel dimensionality 1 required." << EidosTerminate();
-	
+
 	int64_t kernel_dim_a = kernel.dim[0];
-	
+
 	if ((kernel_dim_a < 1) || (kernel_dim_a % 2 == 0))
 		EIDOS_TERMINATION << "ERROR (SpatialMap::Convolve_S1): (internal error) kernel dimensions must be odd." << EidosTerminate();
-	
+
 	int64_t dim_a = grid_size_[0];
 	double *new_values = (double *)malloc(dim_a * sizeof(double));
-	
+
 	if (!new_values)
 		EIDOS_TERMINATION << "ERROR (SpatialMap::Convolve_S1): allocation failed; you may need to raise the memory limit for SLiM." << EidosTerminate(nullptr);
-	
-	// this assumes the kernel's dimensions are symmetrical around its center, and relies on rounding (which is guaranteed)
+
 	int64_t kernel_a_offset = -(kernel_dim_a / 2);
 	double *kernel_values = kernel.values_;
 	double *new_values_ptr = new_values;
-	
-	// FIXME: TO BE PARALLELIZED
+
 	for (int64_t a = 0; a < dim_a; ++a)
 	{
 		double coverage = (!periodic_a_ && ((a == 0) || (a == dim_a - 1))) ? 0.5 : 1.0;
-		
-		// calculate the kernel's effect at point (a)
+
 		double kernel_total = 0.0;
 		double conv_total = 0.0;
-		
-		for (int64_t kernel_a = 0; kernel_a < kernel_dim_a; kernel_a++)
+
+		// Calculate valid kernel range for non-periodic boundaries
+		int64_t kernel_a_first = 0;
+		int64_t kernel_a_last = kernel_dim_a - 1;
+
+		if (!periodic_a_)
 		{
-			int64_t conv_a = a + kernel_a + kernel_a_offset;
-			
-			// clip/wrap to bounds
-			if ((conv_a < 0) || (conv_a >= dim_a))
+			// Clamp to valid range
+			kernel_a_first = std::max((int64_t)0, -(a + kernel_a_offset));
+			kernel_a_last = std::min(kernel_dim_a - 1, (dim_a - 1) - (a + kernel_a_offset));
+		}
+
+		// Check if kernel range is fully within bounds (can use SIMD)
+		int64_t conv_a_first = a + kernel_a_first + kernel_a_offset;
+		int64_t conv_a_last = a + kernel_a_last + kernel_a_offset;
+
+		if (!periodic_a_ && conv_a_first >= 0 && conv_a_last < dim_a)
+		{
+			// All kernel positions are in bounds - use SIMD
+			int64_t count = kernel_a_last - kernel_a_first + 1;
+			Eidos_SIMD::convolve_dot_product_scaled_float64(
+				&kernel_values[kernel_a_first],
+				&values_[conv_a_first],
+				count, coverage,
+				kernel_total, conv_total);
+		}
+		else
+		{
+			// Handle periodic or edge cases element by element
+			for (int64_t kernel_a = 0; kernel_a < kernel_dim_a; kernel_a++)
 			{
-				if (!periodic_a_)
-					continue;
-				
-				// periodicity: assume the two edges have identical values, skip over the edge value on the opposite side
-				while (conv_a < 0)
-					conv_a += (dim_a - 1);	// move -1 to dim - 2
-				while (conv_a >= dim_a)
-					conv_a -= (dim_a - 1);	// move dim to 1
+				int64_t conv_a = a + kernel_a + kernel_a_offset;
+
+				if ((conv_a < 0) || (conv_a >= dim_a))
+				{
+					if (!periodic_a_)
+						continue;
```

Comment:
Codecov says this line isn't covered, suggesting there's no tests with periodic boundary conditions?

Here's what it looks like on Codecov, for @bhaller's interest:
<img width="1252" height="425" alt="Screenshot From 2025-12-19 10-59-49" src="https://github.com/user-attachments/assets/ef07b6f8-9f0d-4f9a-b5d8-a2644b1ae84d" />

Reply:
yay codecov!

Reply:
let me see if I can write a test to hit that...

Reply:
wait that says that it's not hitting the *not* periodic case?